### PR TITLE
Add safe heap_load and heap_store instructions.

### PR DESCRIPTION
**Pull request by lachlansneff**
This pr starts implementing the safe-subset of cranelift IR specified by #589.

### CODE REVIEW

File: lib/wasm/src/code_translator.rs [hunk 1]
```diff
@@ -982,47 +982,10 @@ fn translate_unreachable_operator(
...
-    } else {
-        (base, offset as i32)
-    }
-}
-
 /// Translate a load instruction.
 fn translate_load<FE: FuncEnvironment + ?Sized>(
     offset: u32,
-    opcode: ir::Opcode,
+    _opcode: ir::Opcode,
```
Comment (sunfishcode) [inline lib/wasm/src/code_translator.rs]: We can just remove this argument.

File: lib/wasm/src/code_translator.rs [hunk 2]
```diff
@@ -1031,21 +994,17 @@ fn translate_load<FE: FuncEnvironment + ?Sized>(
...
-    // alignment immediate says it's aligned, because WebAssembly's immediate
-    // field is just a hint, while Cranelift's aligned flag needs a guarantee.
-    let flags = MemFlags::new();
-    let (load, dfg) = builder
-        .ins()
-        .Load(opcode, result_ty, flags, offset.into(), base);
+    let (load, dfg) =
+        builder
+            .ins()
+            .UnaryHeap(ir::Opcode::HeapLoad, result_ty, heap, offset.into(), addr32);
```
Comment (sunfishcode) [inline lib/wasm/src/code_translator.rs]: Does it work to use `builder.ins().heap_load(...)` here?

File: lib/wasm/src/code_translator.rs [hunk 3]
```diff
@@ -1055,12 +1014,14 @@ fn translate_store<FE: FuncEnvironment + ?Sized>(
...
-    builder
-        .ins()
-        .Store(opcode, val_ty, flags, offset.into(), val, base);
+    builder.ins().BinaryHeap(
+        ir::Opcode::HeapStore,
+        val_ty,
+        heap,
+        offset.into(),
+        val,
+        addr32,
```
Comment (sunfishcode) [inline lib/wasm/src/code_translator.rs]: Ditto, with `heap_store`?

File: lib/codegen/meta-python/base/instructions.py [hunk 4]
```diff
@@ -609,6 +609,50 @@
         """,
         ins=(T, p, Offset), outs=addr)
 
+#
+# Heap accesses for safe-subset of cranelift ir.
```
Comment (sunfishcode) [inline lib/codegen/meta-python/base/instructions.py]: Capitalize Cranelift and IR.

File: lib/codegen/meta-python/base/instructions.py [hunk 5]
```diff
@@ -609,6 +609,50 @@
...
+H = Operand('H', entities.heap)
+HeapOffset = TypeVar('HeapOffset', 'An unsigned heap offset', ints=(32, 64))
+x = Operand('x', Mem, doc='Value to be stored')
+a = Operand('a', Mem, doc='Value loaded')
+p = Operand('p', HeapOffset)
+Offset = Operand('Offset', uimm32, 'Byte offset from element address')
+
+heap_load = Instruction(
+        'heap_load', r"""
+        Safely load from a cranelift heap.
```
Comment (sunfishcode) [inline lib/codegen/meta-python/base/instructions.py]: Capitalize Cranelift.

File: lib/codegen/meta-python/base/instructions.py [hunk 6]
```diff
@@ -609,6 +609,50 @@
...
+           the value at ``p + Offset``.
+        2. If ``p + Offset + sizeof a`` is greater than the heap bound, generate a trap.
+        """,
+        ins=(H, p, Offset), outs=a, can_load=True)
+
+heap_store = Instruction(
+        'heap_store', r"""
+        Store ``x`` to memory at ``p + Offset``.
+
+        Safely store in a cranelift heap.
```
Comment (sunfishcode) [inline lib/codegen/meta-python/base/instructions.py]: ditto

**Comment by lachlansneff** [comment]
We need to think more about how to generalize `heap_load` and `heap_store` for all kinds of those operations (e.g. different sizes).

File: lib/codegen/meta-python/base/legalize.py [hunk 7]
```diff
@@ -94,6 +94,10 @@
 expand.custom_legalize(insts.stack_load, 'expand_stack_load')
 expand.custom_legalize(insts.stack_store, 'expand_stack_store')
 
+# CUstom expansions for heap_load/heap_store
```
Comment (bjorn3) [inline lib/codegen/meta-python/base/legalize.py]:
```suggestion
# Custom expansions for heap_load/heap_store
```

**Pull request closed without merging**